File: binding-java/runtime/src/main/java/org/apache/etch/bindings/java/transport/DefaultDeliveryService.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.transport;

import org.apache.etch.bindings.java.msg.Message;
import org.apache.etch.bindings.java.msg.Type;
import org.apache.etch.bindings.java.support.DeliveryService;
import org.apache.etch.bindings.java.support.Mailbox;
import org.apache.etch.util.Monitor;
import org.apache.etch.util.Resources;
import org.apache.etch.util.TimeoutException;
import org.apache.etch.util.URL;
import org.apache.etch.util.core.Who;


/**
 * Default implementation of DeliveryService.
 */
public class DefaultDeliveryService implements DeliveryService
{
	
	/**
	 * Parameter to Globally Disable Timeout.
	 */
	public final static String DISABLE_TIMEOUT = "DefaultDeliveryService.disableTimeout";
	
	
	private boolean disableTimeout = false;
	
	/**
	 * @param transport
	 * @param uri
	 * @param resources
	 */
	public DefaultDeliveryService( MailboxManager transport, String uri,
		Resources resources )
	{
		this( transport, new URL( uri ), resources );
	}

	/**
	 * @param transport
	 * @param uri
	 * @param resources
	 */
	public DefaultDeliveryService( MailboxManager transport, URL uri,
		Resources resources )
	{
		this.transport = transport;
		transport.setSession( this );
		disableTimeout = uri.getBooleanTerm( DISABLE_TIMEOUT , false );
	}
	
	private final MailboxManager transport;
	
	/**
	 * @return the transport.
	 */
	public MailboxManager getTransport()
	{
		return transport;
	}
	
	public SessionMessage getSession()
	{
		return session;
	}
	
	public void setSession( SessionMessage session )
	{
		if (this.session != null)
			throw new UnsupportedOperationException( "only one stub for now" );
		this.session = session;
	}
	
	private SessionMessage session;

	public Object sessionQuery( Object query ) throws Exception
	{
		return session.sessionQuery( query );
	}

	public void sessionControl( Object control, Object value ) throws Exception
	{
		session.sessionControl( control, value );
	}

	public void sessionNotify( Object event ) throws Exception
	{
		if (event == UP)
			status.set( UP );
		else if (event == DOWN)
			status.set( DOWN );
		
		session.sessionNotify( event );
	}
	
	private final Monitor<String> status = new Monitor<String>( "session status" );

	public boolean sessionMessage( Who sender, Message msg ) throws Exception
	{
		return session.sessionMessage( sender, msg );
	}

	@Override
	public String toString()
	{
		return transport.toString();
	}
	
	private void waitUp( int maxDelay ) throws Exception
	{
		status.waitUntilEq( UP, maxDelay );
	}
	
	private void waitDown( int maxDelay ) throws Exception
	{
		status.waitUntilEq( DOWN, maxDelay );
	}

	public void transportMessage( Who recipient, Message msg ) throws Exception
	{
		transport.transportMessage( recipient, msg );
	}
	
	public Object transportQuery( Object query ) throws Exception
	{
		if (query.getClass() == WaitUp.class)
		{
			waitUp( ((WaitUp) query).maxDelay );
			return null;
		}
		else if (query.getClass() == WaitDown.class)
		{
			waitDown( ((WaitDown) query).maxDelay );
			return null;
		}
		else
		{
			return transport.transportQuery( query );
		}
	}
	
	public void transportControl( Object control, Object value ) throws Exception
	{
		if (control == START_AND_WAIT_UP)
		{
			transport.transportControl( START, null );
			waitUp( (Integer) value );
		}
		else if (control == STOP_AND_WAIT_DOWN)
		{
			transport.transportControl( STOP, null );
			waitDown( (Integer) value );
		}
		else
		{
			transport.transportControl( control, value );
		}
	}

	public void transportNotify( Object event ) throws Exception
	{
		transport.transportNotify( event );
	}

	public Mailbox begincall( Message msg )
	{
		try
		{
			return transport.transportCall( null, msg );
		}
		catch ( Exception e )
		{
			throw new RuntimeException( "unexpected exception sending message", e );
		}
	}

	public Object endcall( Mailbox mb, Type responseType ) throws Exception
	{
		try
		{
			int timeout = disableTimeout ? 0 : responseType.getTimeout();
			Mailbox.Element mbe = mb.read( timeout);
			if (mbe == null)
				throw new TimeoutException( "timeout waiting for "+responseType );
			Message rmsg = mbe.msg;
			rmsg.checkType( responseType );
			Object r = rmsg.get( responseType.getResponseField() );
			if (r instanceof Exception)
			{
				Exception e = (Exception) r;
				e.fillInStackTrace();
				throw e;
			}
			return r;
		}
		finally
		{
			mb.closeRead();
		}
	}
}

File: binding-python/runtime/src/main/python/etch/binding/transport/DefaultDeliveryService.py
"""
#  Licensed to the Apache Software Foundation (ASF) under one   *
#  or more contributor license agreements.  See the NOTICE file *
#  distributed with this work for additional information        *
#  regarding copyright ownership.  The ASF licenses this file   *
#  to you under the Apache License, Version 2.0 (the            *
#  "License"); you may not use this file except in compliance   *
#  with the License.  You may obtain a copy of the License at   *
#                                                               *
#    http://www.apache.org/licenses/LICENSE-2.0                 *
#                                                               *
#  Unless required by applicable law or agreed to in writing,   *
#  software distributed under the License is distributed on an  *
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
#  KIND, either express or implied.  See the License for the    *
#  specific language governing permissions and limitations      *
#  under the License. 
"""
from __future__ import absolute_import
from ..msg.Message import *
from ..msg.Type import *
from ..support.DeliveryService import *
from ..support.Mailbox import *
from ...util.Monitor import *
from ...util.Resources import *
from ...util.TimeoutException import *
from ...util.URL import *
from ...util.core.Who import *

 
class DefaultDeliveryService(DeliveryService):
    """
    Default implementation of DeliveryService
    """
    
    DISABLE_TIMEOUT = "DefaultDeliveryService.disableTimeout"
    """Parameter to Globally Disable Timeout."""
    
    def __init__(self, transport, uri, resources):
        """
        @param transport
        @param uri
        @param resources
        """
        self.__transport = transport
        self.__session = None
        self.__status  = Monitor("session status")
        u = URL(uri)
        self.__transport.setSession(self)
        self.__disableTimeout = u.getBooleanTerm(self.DISABLE_TIMEOUT, False)

    def getTransport(self):
        """
        @return the transport
        """
        return self.__transport
    
    def getSession(self):
        return self.__session
        
    def sessionQuery(self, query):
        return self.__session.sessionQuery(query)
    
    def sessionControl(self, control, value):
        self.__sessionControl(control, value)
    
    def sessionNotify(self, event):
        if event == self.UP:
            self.__status.set(self.UP)
        elif event == self.DOWN:
            self.__status.set(self.DOWN)
        
        self.__session.sessionNotify(event)
    
    def sessionMessage(self, sender, msg):
        return self.__session.sessionMessage(sender, msg)
    
    def __repr__(self):
        return self.__transport.__repr__()
    
    def waitUp(self, maxDelay):
        self.__status.waitUntilEq(self.UP, maxDelay)
    
    def waitDown(self, maxDelay):
        self.__status.waitUntilEq(self.DOWN, maxDelay)
    
    def transportMessage(self, recipient, msg):
        self.__transport.transportMessage(recipient, msg)
    
    def transportQuery(self, query):
        #TODO - where are WaitUp and WaitDown defined?
        if query.__class__ == WaitUp:
            self.waitUp(query.maxDelay)
            return None
        elif query.__class__ == WaitDown:
            self.waitDown(query.maxDelay)
            return None
        else:
            return self.__transport.transportQuery(query)
    
    def transportControl(self, control, value):
        if control == self.START_AND_WAIT_UP:
            self.__transport.transportControl(self.START, None)
            self.waitUp(value)
        elif control == self.STOP_AND_WAIT_DOWN:
            self.__transport.transportControl(self.STOP, None)
            self.waitDown(value)
        else:
            self.__transport.transportControl(control, value)
    
    def transportNotify(self, event):
        self.__transport.transportNotify(event)

    def beginCall(self, msg):
        try:
            return self.__transport.transportCall(None, msg)
        except:
            # TODO - more detail in raise?
            raise Exception, "unexpected exception sending message"
    
    def endCall(self, mb, responseType):
        try:
            timeout = 0
            if not self.__disableTimeout:
                timeout = responseType.getTimeout()
            
            mbe = mb.read(timeout)
            if mbe == None:
                raise TimeoutException, "timeout waiting for %s" % repr(responseType)
            
            rmsg = mbe.msg
            rmsg.checkType(responseType)
            r = rmsg.get(responseType.getResponseField())
            if isinstance(r, Exception):
                raise r
            return r
        finally:
            mb.closeDelivery()
    
    
    
